eip-0008.md

```diff
@@ -0,0 +1,479 @@
+
+If a transaction "runs out of gas" mid-execution, then all execution reverts, but the
+transaction is nevertheless valid, and the only effect of the transaction is to transfer
+the entire sum `gasLimit * gasPrice` to the miner.
+
+JIT costing has the following pros and cons.
+
+##### Pros
+1) Additional Economic incentive for miners to execute contract. The amount of purchase
+from sender's account `(gasLimit - gasRem) * gasPrice` is payed to the miner, which
+reflects the actual execution costs.
+2) Simple implementation. Cost Table for all VM commands, accumulate the cost along an
+execution trace.
+3) Accurate approximation. Only actually executed operations are summed up. The execution
+trace depends on the current state of the contract's data (i.e. current context) and
+thus also the accumulated cost (i.e. `(gasLimit - gasRem)` value).
+  
+##### Cons
+1) Transaction sender will loose funds in case of exceeding `gasLimit`. In a
+turing-complete language of Ethereum VM (byte code instruction set) it may be hard to
+choose appropriate `gasLimit` to associate with the transaction. This is in particular
+due to the point 3) in the Pros list. As a result the transaction will be considered
+invalid if actual (and unpredictably big) gas consumption exceeds `gasLimit`, in which
+case the sender looses the funds.
+2) The actual complexity of contracts may be limited by `gasPrice`.
+Since all computations require gas, the complexity of contracts becomes limited by not
+the language (which is powerful) and not the actual computation expenses of the miners, 
+but by the current `gasPrice` asked by the miners. You would not want to spend more of your
+funds for the gas required to execute a complex contract, than the amounts of funds you operate
+with (i.e. sending $10 and pay $100 for the gas doesn't have economic sense).
+
+### Ergo Transaction Validation With JIT Costing
+
+Here we describe an extension to the existing Ergo protocol, which allows to prevent execution
+of scripts beyond allotted limits. The extension doesn't require changes in the
+serialization format. Moreover, simple _transfer transactions_ are processed the same way so
+that 1) most existing applications (like exchanges) will not require any changes and 2)
+the new protocol doesn't incur additional overhead for simple transfer transactions.
+The cost calculation of ErgoTrees is implemented as it is defined in [the next
+section](#jit-costing-of-ergotree).
+
+On a high level, the new protocol for transaction creation, signing and validation is
+described as a series of steps. The anatomy of the transactions which are created is shown
+in the figure below.
+
+ 1) A transaction sender performs cost estimation as part of transaction signing.
+Transaction signing involves: 1) execution of the scripts for all the input boxes and 2)
+execution of sigma protocols to generate proofs. This happens using the current context of
```

Comment:
> 1) A transaction sender performs cost estimation as part of transaction signing.
> Transaction signing involves: 1) execution of the scripts for all the input boxes and 2)
> execution of sigma protocols to generate proofs.


Doesn't this incur a really high barrier of entry for light clients? This would mean they all need to have support for executing the sigma protocols which is not trivial to have implemented and is actually quite a bit to ask for implementors. 

This would likely also stop more specialized light-clients who only have a few key contracts they support with various registers they change across txs to perform actions.

Reply:
To be honest, supporting full sigma interpreter is not hard problem at all, the language is simple and the interpreter is straightforward. 
We have at least two options to have sigma on ALL platforms (Rust and Dart based).
We are doing Rust, and after some time this will not be an issue.
So, yes, light clients will be able to sign transaction in normal way and estimate costs.
I don't see this is a big deal.

The hardest part of existing interpreter is actually AOT costing which is not necessary on the client side.